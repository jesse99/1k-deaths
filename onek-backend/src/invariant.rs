use super::*;

// TODO: do fast invariants always in debug, check limited number of cells
// TODO: do slow invariants if an option is set, check all cells
#[cfg(debug_assertions)]
pub fn invariant(game: &Game) {
    if game.invariants {
        for (loc, oids) in game.level.iter() {
            check_cell(game, *loc, oids);
        }
    } else {
        // Check up to an 8x8 grid around the player, ignoring LOS.
        const DELTA: i32 = 4;
        for dy in -DELTA..=DELTA {
            for dx in -DELTA..=DELTA {
                let loc = Point::new(game.player_loc.x + dx, game.player_loc.y + dy);
                if let Some(oids) = game.level.get(&loc) {
                    check_cell(game, loc, oids);
                }
            }
        }
    }
}

#[cfg(debug_assertions)]
pub fn check_cell(game: &Game, loc: Point, oids: &Vec<Oid>) {
    for (i, oid) in oids.iter().enumerate() {
        let object = game.objects.get(oid).expect(&format!("No object for {oid} at {loc}"));
        if i == 0 {
            // first object in a cell should be terrain
            assert!(
                object.contains_key("blocks_los"),
                "first object at {loc} is missing blocks_Los"
            );
            assert!(object.contains_key("color"), "first object at {loc} is missing color");
            assert!(
                object.contains_key("back_color"),
                "first object at {loc} is missing back_color"
            );
        }

        // all objects should have tag, oid, description, symbol
        assert!(object.contains_key("tag"), "first object at {loc} is missing tag");
        assert!(object.contains_key("oid"), "first object at {loc} is missing oid");
        assert!(
            object.contains_key("description"),
            "first object at {loc} is missing description"
        );
        assert!(object.contains_key("symbol"), "first object at {loc} is missing symbol");

        // oids should be consistent
        let value = object.get("oid").unwrap().to_oid();
        assert!(value == oid, "object has oid {value} but level has {oid}");

        // player should only be at player_loc
        if *oid == PLAYER_OID {
            assert!(
                loc == game.player_loc,
                "player should be at {} not {loc}",
                game.player_loc,
            );
        }
    }
}
