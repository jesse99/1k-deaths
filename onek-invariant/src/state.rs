#[cfg(test)]
use ipmpsc::{Receiver, Sender, SharedRingBuffer};
#[cfg(test)]
use onek_types::*;
#[cfg(test)]
use std::time::Duration;

#[cfg(test)] // TODO: get rid of this once we start listening to State updates
pub struct State {
    pub tx: ipmpsc::Sender,
    pub rx: ipmpsc::Receiver,
    pub rx_name: ChannelName,
}

#[cfg(test)]
impl State {
    pub fn new(map: &str) -> State {
        let tx = match SharedRingBuffer::open("/tmp/state-sink") {
            Ok(buffer) => Sender::new(buffer),
            Err(err) => panic!("error opening state-sink: {err:?}"),
        };

        // Note that we have to do this early because Reset will zap the RegisterForQuery below.
        let mesg = StateMessages::Mutate(StateMutators::Reset(map.to_string()));
        let result = tx.send(&mesg);
        assert!(!result.is_err(), "error sending Reset to State: {result:?}");

        let name = "/tmp/tester-sink";
        let rx = match SharedRingBuffer::create(name, 32 * 1024) {
            Ok(buffer) => Receiver::new(buffer),
            Err(err) => panic!("error opening tester-sink: {err:?}"),
        };

        let mesg = StateMessages::RegisterForQuery(ChannelName::new(name));
        let result = tx.send(&mesg);
        assert!(!result.is_err(), "error sending RegisterForQuery to State: {result:?}");

        State {
            tx,
            rx,
            rx_name: ChannelName::new(name),
        }
    }

    // pub fn send_mutate(&self, mutate: StateMutators) {
    //     let mesg = StateMessages::Mutate(mutate.clone());
    //     let result = self.tx.send(&mesg);
    //     assert!(!result.is_err(), "error sending {mutate:?} to State: {result:?}")
    // }

    pub fn get_player_view(&self) -> View {
        let query = StateQueries::PlayerView(self.rx_name.clone());
        let response = self.send_query(query);
        match response {
            StateResponse::Map(map) => map,
            _ => panic!("Expected View from PlayerView query not {response}"),
        }
    }

    pub fn get_player_loc(&self) -> Point {
        let query = StateQueries::PlayerLoc(self.rx_name.clone());
        let response = self.send_query(query);
        match response {
            StateResponse::Location(loc) => loc,
            _ => panic!("Expected Point from PlayerLoc query not {response}"),
        }
    }

    pub fn send_query(&self, query: StateQueries) -> StateResponse {
        let mesg = StateMessages::Query(query.clone());
        let result = self.tx.send(&mesg);
        assert!(!result.is_err(), "error sending {query:?} to State: {result:?}");

        let result = self.rx.recv_timeout(Duration::from_millis(100));
        assert!(!result.is_err(), "error receiving from State: {result:?}");

        let option = result.unwrap();
        assert!(option.is_some(), "timed out receiving {query:?} from State");

        option.unwrap()
    }
}
