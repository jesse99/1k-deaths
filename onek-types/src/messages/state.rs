use crate::*;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fmt;

#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum NoteKind {
    Error,
}

/// These are in-game messages for the player, e.g. combat results or status messages.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Note {
    pub text: String,
    pub kind: NoteKind,
}

#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
pub enum Terrain {
    Dirt,
    Wall,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Cell {
    pub terrain: Terrain,
    pub objects: Vec<Oid>,
    pub character: Option<Oid>,
}

/// Represents a portion of a level. Typically cells visible to a character.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct View {
    pub cells: HashMap<Point, Cell>,

    /// top_left and bottom_right are the smallest rectangle enclosing all the locations.
    pub top_left: Point,
    pub bottom_right: Point,
}

impl View {
    pub fn new() -> View {
        View {
            cells: HashMap::new(),
            top_left: Point::origin(),
            bottom_right: Point::origin(),
        }
    }

    pub fn insert(&mut self, loc: Point, cell: Cell) {
        if self.cells.is_empty() {
            self.top_left = loc;
            self.bottom_right = loc;
        } else {
            if loc.x < self.top_left.x {
                self.top_left.x = loc.x;
            }
            if loc.y < self.top_left.y {
                self.top_left.y = loc.y;
            }

            if loc.x > self.bottom_right.x {
                self.bottom_right.x = loc.x;
            }
            if loc.y > self.bottom_right.y {
                self.bottom_right.y = loc.y;
            }
        }

        self.cells.insert(loc, cell);
    }

    pub fn size(&self) -> Size {
        self.bottom_right - self.top_left
    }
}

/// These take the name of a channel to send a [`StateResponse`] to.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum StateQueries {
    CellAt(ChannelName, Point),
    Notes(ChannelName, usize),
    PlayerView(ChannelName),
    PlayerLoc(ChannelName),
}

/// These update internal state and then send a StateResponse.Updated message to services
/// that used RegisterForUpdate. These return StateResponse.Mutated.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum StateMutators {
    /// Read transactions allow services to get a consistent view of state: any mutation
    /// requests that come in during the transaction are deferred until the transaction
    /// ends. The String argument is an ID used to match up the begin and end of transactions.
    BeginReadTransaction(String),
    EndReadTransaction(String),

    AddNote(Note),
    MovePlayer(Point),
    Reset(String), // could include an arg to map weird chars to some sort of object enum
}

/// Messages that the state service receives.
#[derive(Debug, Serialize, Deserialize)]
pub enum StateMessages {
    Mutate(StateMutators),
    Query(StateQueries),
    RegisterForQuery(ChannelName),
    RegisterForUpdate(ChannelName),
}

/// Messages that the state service sends to other services.
#[derive(Debug, Serialize, Deserialize)]
pub enum StateResponse {
    // TODO: ready to move should include all objects that are ready
    Cell(Cell),
    Location(Point),
    Map(View),
    Notes(Vec<Note>),
    Updated(EditCount),
}

mod display_impl {
    use super::*;

    impl fmt::Display for StateMessages {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(f, "{:?}", self)
        }
    }

    impl fmt::Display for StateResponse {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(f, "{:?}", self)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_view() {
        #[rustfmt::skip]
        let mut view = View::new();
        let cell = Cell {
            terrain: Terrain::Dirt,
            objects: Vec::new(),
            character: None,
        };

        view.insert(Point::new(10, 10), cell.clone());
        assert_eq!(view.top_left, Point::new(10, 10));
        assert_eq!(view.bottom_right, Point::new(10, 10));

        view.insert(Point::new(15, 15), cell.clone());
        assert_eq!(view.top_left, Point::new(10, 10));
        assert_eq!(view.bottom_right, Point::new(15, 15));

        view.insert(Point::new(5, 12), cell.clone());
        assert_eq!(view.top_left, Point::new(5, 10));
        assert_eq!(view.bottom_right, Point::new(15, 15));

        view.insert(Point::new(12, 20), cell.clone());
        assert_eq!(view.top_left, Point::new(5, 10));
        assert_eq!(view.bottom_right, Point::new(15, 20));
    }
}
