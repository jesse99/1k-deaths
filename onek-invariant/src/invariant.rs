#[cfg(test)]
use super::state::*;
#[cfg(test)]
use onek_types::PLAYER_ID;

// TODO: probably should check the entire map
// expensive but this is in its own process and we don't have to
// process every update
// but that could blow the message queue!
#[cfg(test)]
pub fn invariant(state: &State) {
    state.begin_read_transaction("invariant".to_string());

    let view = state.get_player_view();
    assert!(!view.cells.is_empty(), "need at least the cell the player is at");

    let loc = state.get_player_loc();
    let cell = view.cells.get(&loc);
    assert!(cell.is_some(), "expected a cell for the player at {loc}");

    let ch = cell.unwrap().character;
    assert!(ch.is_some(), "expected a character for the player at {loc}");

    let oid = ch.unwrap();
    assert!(oid == PLAYER_ID, "expected player at {loc} not {oid}");

    state.end_read_transaction("invariant".to_string());
}
