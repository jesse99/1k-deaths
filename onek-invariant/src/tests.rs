#[cfg(test)]
use super::invariant::*;
#[cfg(test)]
use super::state::*;
#[cfg(test)]
use onek_types::*;

#[cfg(test)]
trait ToSnapshot {
    fn to_snapshot(&self, state: &State) -> String;
}

#[cfg(test)]
fn terrain_to_char(terrain: Terrain) -> char {
    match terrain {
        Terrain::Dirt => ' ',
        Terrain::Wall => '#',
    }
}

#[cfg(test)]
impl ToSnapshot for StateResponse {
    fn to_snapshot(&self, state: &State) -> String {
        match self {
            StateResponse::Map(map) => map.to_snapshot(state),
            _ => panic!("snapshots are not supported for {self}"),
        }
    }
}

#[cfg(test)]
impl ToSnapshot for View {
    fn to_snapshot(&self, _test: &State) -> String {
        let mut result = String::with_capacity(200);
        for y in self.top_left.y..=self.top_left.y + self.size().height {
            for x in self.top_left.x..=self.top_left.x + self.size().width {
                let loc = Point::new(x, y);
                match self.cells.get(&loc) {
                    Some(cell) => {
                        if cell.character.unwrap_or(NULL_ID) == PLAYER_ID {
                            result.push('@');
                        } else {
                            result.push(terrain_to_char(cell.terrain));
                        }
                    }
                    None => result.push(' '),
                }
            }
            result.push('\n');
        }
        // At some point will need to use tx to include details about objects.
        result
    }
}

// Mutators
#[cfg(test)]
impl State {}

#[test]
fn test_from_str() {
    let state = State::new(
        "###\n\
             #@#\n\
             ###",
    );

    invariant(&state);

    let view = state.get_player_view();
    insta::assert_display_snapshot!(view.to_snapshot(&state));
}
