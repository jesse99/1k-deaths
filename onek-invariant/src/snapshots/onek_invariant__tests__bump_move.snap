---
source: onek-invariant/src/tests.rs
expression: view.to_snapshot(&state)
---
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????####????????
????????# @#????????
????????####????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????
????????????????????

